tests/neg/t4815.scala

```diff
@@ -0,0 +1,24 @@
+class Test {
+  def foo = {
+    final def bar = 1 // error: local def may not be final
```

Comment:
`scalac` rejects this. I think we should reject it as well

Reply:
I'm not sure I understand. We do reject this. This is what this PR does.

Reply:
Sorry, it was meant for the line below (i.e. `final val v = 42`)

Reply:
Ah, I see. There's at least one test that uses this feature (constant value expression):
- https://github.com/lampepfl/dotty/blob/946ba3f2d938b54e316749aef1c2ff117ecdb325/tests/run/tasty-eval/quoted_2.scala#L8
- Additionally, it seem to be explicitly allowed by the spec: https://scala-lang.org/files/archive/spec/2.12/04-basic-declarations-and-definitions.html#value-declarations-and-definitions

```
A constant value definition is of the form

final val x = e

where e is a constant expression. The final modifier must be present and no type annotation may be given. References to the constant value x are themselves treated as constant expressions; in the generated code they are replaced by the definition's right-hand side e.
```

Reply:
This part of the spec only specifies what a constant value is, not where it is allowed. Looking at the [syntax part of the spec](https://scala-lang.org/files/archive/spec/2.12/13-syntax-summary.html), it doesn't seems to be allowed
```
  BlockStat         ::=  Import
                      |  {Annotation} [‘implicit’ | ‘lazy’] Def
                      |  {Annotation} {LocalModifier} TmplDef
                      |  Expr1
  Def               ::=  PatVarDef
                      |  ‘def’ FunDef
                      |  ‘type’ {nl} TypeDef
                      |  TmplDef
  PatVarDef         ::=  ‘val’ PatDef
                      |  ‘var’ VarDef
```